[42] CMake Alias Targets

RFRIEDM-Trimble opened the pull request:
- Add CMake ALIAS targets for all current libraries
- Consume the new alias targets in all calls to target_link_libraries
- Update gtest to be found using either find_package or the fetchcontent method
- Add missing alias to gtest for pre cmake 3.20 support to be unified
- Solves #42

Signed-off-by: Ryan Friedman <ryan_friedman@trimble.com>

RFRIEDM-Trimble commented:
For due diligence, complete the following before merge:

- [ ] Run a build with cmake 3.16 and gtest installed in the system with apt
- [ ] Run a build with cmake 3.16 and gtest NOT installed in the system so it uses the fetchcontent method
- [ ] Run a build with cmake 3.23 and gtest installed in the system with apt
- [ ] Run a build with cmake 3.23 and gtest NOT installed in the system so it uses the fetchcontent method

The point is to check both system install and fetchcontent methods work for before/after cmake 3.20 which has different behavior

RFRIEDM-Trimble commented:
> For due diligence, complete the following before merge:
> 
> * [ ]  Run a build with cmake 3.16 and gtest installed in the system with apt
> * [ ]  Run a build with cmake 3.16 and gtest NOT installed in the system so it uses the fetchcontent method
> * [ ]  Run a build with cmake 3.23 and gtest installed in the system with apt
> * [ ]  Run a build with cmake 3.23 and gtest NOT installed in the system so it uses the fetchcontent method
> 
> The point is to check both system install and fetchcontent methods work for before/after cmake 3.20 which has different behavior. I can run these tests in a local docker environment, but need some time to do it.

Test 3 failing for unrelated compilation errors in alpine Linux. I will take a workaround to try to install different versions of the compiler to more match the tested host build system you use.
```
docker run -v $PWD:/isobus -it alpine:edge
apk add --no-cache cmake g++ make git linux-headers
rm -r build && cmake -S . -B build && cmake --build build
```
```
-- The CXX compiler identification is GNU 12.2.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- The C compiler identification is GNU 12.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Could NOT find Python (missing: Python_EXECUTABLE Interpreter) 
-- Configuring done
-- Generating done
...
[ 63%] Building CXX object socket_can/CMakeFiles/SocketCANInterface.dir/src/socket_can_interface.cpp.o
In file included from /usr/include/net/if.h:27,
                 from /ISO11783-CAN-Stack/socket_can/src/socket_can_interface.cpp:15:
/ISO11783-CAN-Stack/socket_can/src/socket_can_interface.cpp: In member function 'bool CANHardwareInterface::SocketCANFrameHandler::read_frame(isobus::HardwareInterfaceCANFrame&)':
/ISO11783-CAN-Stack/socket_can/src/socket_can_interface.cpp:144:49: error: invalid application of 'sizeof' to incomplete type 'CANHardwareInterface::SocketCANFrameHandler::read_frame(isobus::HardwareInterfaceCANFrame&)::timeval'
  144 |                 char lControlMessage[CMSG_SPACE(sizeof(struct timeval) + (3 * sizeof(struct timespec)) + sizeof(std::uint32_t))];
      |                                                 ^~~~~~~~~~~~~~~~~~~~~~
/ISO11783-CAN-Stack/socket_can/src/socket_can_interface.cpp:150:40: error: 'lControlMessage' was not declared in this scope
  150 |                 message.msg_control = &lControlMessage;
      |                                        ^~~~~~~~~~~~~~~
/ISO11783-CAN-Stack/socket_can/src/socket_can_interface.cpp:184:120: error: invalid use of incomplete type 'struct CANHardwareInterface::SocketCANFrameHandler::read_frame(isobus::HardwareInterfaceCANFrame&)::timeval'
  184 |                                                                 canFrame.timestamp_us = static_cast<std::uint64_t>(time->tv_usec) + (static_cast<std::uint64_t>(time->tv_sec) * 1000000);
      |                                                                                                                        ^~
/ISO11783-CAN-Stack/socket_can/src/socket_can_interface.cpp:144:63: note: forward declaration of 'struct CANHardwareInterface::SocketCANFrameHandler::read_frame(isobus::HardwareInterfaceCANFrame&)::timeval'
  144 |                 char lControlMessage[CMSG_SPACE(sizeof(struct timeval) + (3 * sizeof(struct timespec)) + sizeof(std::uint32_t))];
      |                                                               ^~~~~~~
/ISO11783-CAN-Stack/socket_can/src/socket_can_interface.cpp:184:165: error: invalid use of incomplete type 'struct CANHardwareInterface::SocketCANFrameHandler::read_frame(isobus::HardwareInterfaceCANFrame&)::timeval'
  184 |                                                                 canFrame.timestamp_us = static_cast<std::uint64_t>(time->tv_usec) + (static_cast<std::uint64_t>(time->tv_sec) * 1000000);
      |                                                                                                                                                                     ^~
/ISO11783-CAN-Stack/socket_can/src/socket_can_interface.cpp:144:63: note: forward declaration of 'struct CANHardwareInterface::SocketCANFrameHandler::read_frame(isobus::HardwareInterfaceCANFrame&)::timeval'
  144 |                 char lControlMessage[CMSG_SPACE(sizeof(struct timeval) + (3 * sizeof(struct timespec)) + sizeof(std::uint32_t))];
      |                                                               ^~~~~~~
make[2]: *** [socket_can/CMakeFiles/SocketCANInterface.dir/build.make:76: socket_can/CMakeFiles/SocketCANInterface.dir/src/socket_can_interface.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1100: socket_can/CMakeFiles/SocketCANInterface.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```

ad3154 commented:
> Test 3 failing for unrelated compilation errors in alpine Linux. I will take a workaround to try to install different versions of the compiler to more match the tested host build system you use.

Added issue #44 for Alpine compilation issue

RFRIEDM-Trimble commented:
Rebased PR on `main` to fix the compilation error. All tests pass except the first one. Looking into a fix. 

```
CMake Error at CMakeLists.txt:77 (add_executable):
  Target "unit_tests" links to target "GTest::gtest_main" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?


CMake Error at CMakeLists.txt:77 (add_executable):
  Target "unit_tests" links to target "GTest::gtest_main" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
  ```
  
  On ubuntu 20.04, it uses gtest `1.10.0`, which doesn't create the imported target as needed here with GLOBAL scope. 
  
  I implemented a fix in here: 0bd924f626b6cbaa80740f787bc27d24bc6b6928
  It follows this reference: https://stackoverflow.com/questions/45401212/how-to-make-imported-target-global-afterwards

RFRIEDM-Trimble commented:
If you aren't happy with the approach here, perhaps an alternative is just to use generators, linking if the target exists like so: https://stackoverflow.com/questions/66744748/using-cmake-generator-expressions-to-link-if-target-exists

ad3154 commented on socket_can/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+cmake_minimum_required(VERSION 3.16)
```
The CMakeLists.txt in socket_can directory can be removed as that folder was removed here: #49

RFRIEDM-Trimble replied:
Removed and rebased.

RFRIEDM-Trimble commented:
Seems like tutorials also have usage instructions with calls to target_link_libraries. Did you want those cleaned up here too?

ad3154 commented:
> Seems like tutorials also have usage instructions with calls to target_link_libraries. Did you want those cleaned up here too?

Up to you, If you're willing to do that it would be excellent! Otherwise I can make a follow-up PR later. I have to redeploy the site anyways if it changes so it's same-ish amount of work for me either way

RFRIEDM-Trimble commented:
> > Seems like tutorials also have usage instructions with calls to target_link_libraries. Did you want those cleaned up here too?
> 
> Up to you, If you're willing to do that it would be excellent! Otherwise I can make a follow-up PR later. I have to redeploy the site anyways if it changes so it's same-ish amount of work for me either way

I'd prefer a follow up, I'll create a ticket. There's other stuff that't not modern cmake in the tutorials like the linkage to threads that can be cleaned up.
